Add a functional background fetch backend

youennf opened the pull request:
#### 1767a09f5be8c26c8b4e95a11cfe9af937a0f114
<pre>
Add a functional background fetch backend
<a href="https://bugs.webkit.org/show_bug.cgi?id=252657">https://bugs.webkit.org/show_bug.cgi?id=252657</a>
rdar://problem/105720611

Reviewed by Chris Dumez and Sihui Liu.

Implement background fetch backend as follows:
- BackgroundFetchEngine is the class that is used by SWServer to implement the background fetch functionality
- BackgroundFetch represents a back ground fetch and has BackgroundFetch::Record(s).
- BackgroundFetchRecordLoader and its BackgroundFetchLoad implementation is about the actual network load.
- BackgroundFetchStore and its WebKit implementation BackgroundFetchStoreImpl allow storing background fetches.
BackgroundFetchStoreImpl is currently an empty shell.

Covered by rebased tests.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/background-fetch/abort.https.window-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/background-fetch/content-security-policy.https.window-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/background-fetch/fetch-uploads.https.window-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/background-fetch/fetch.https.window-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/background-fetch/get-ids.https.window-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/background-fetch/get.https.window-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/background-fetch/match.https.window-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/background-fetch/mixed-content-and-allowed-schemes.https.window-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/background-fetch/port-blocking.https.window-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/background-fetch/update-ui.https.window-expected.txt:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/workers/service/ServiceWorkerRegistrationKey.h:
* Source/WebCore/workers/service/background-fetch/BackgroundFetch.cpp: Added.
(WebCore::BackgroundFetch::BackgroundFetch):
(WebCore::BackgroundFetch::~BackgroundFetch):
(WebCore::BackgroundFetch::information const):
(WebCore::BackgroundFetch::match):
(WebCore::BackgroundFetch::abort):
(WebCore::BackgroundFetch::perform):
(WebCore::BackgroundFetch::storeResponse):
(WebCore::BackgroundFetch::storeResponseBodyChunk):
(WebCore::BackgroundFetch::didSendData):
(WebCore::BackgroundFetch::didFinishRecord):
(WebCore::BackgroundFetch::handleStoreResult):
(WebCore::BackgroundFetch::recordIsCompleted):
(WebCore::BackgroundFetch::updateBackgroundFetchStatus):
(WebCore::BackgroundFetch::unsetRecordsAvailableFlag):
(WebCore::BackgroundFetch::Record::Record):
(WebCore::BackgroundFetch::Record::~Record):
(WebCore::BackgroundFetch::Record::isMatching const):
(WebCore::BackgroundFetch::Record::information const):
(WebCore::BackgroundFetch::Record::complete):
(WebCore::BackgroundFetch::Record::abort):
(WebCore::BackgroundFetch::Record::didSendData):
(WebCore::BackgroundFetch::Record::didReceiveResponse):
(WebCore::BackgroundFetch::Record::didReceiveResponseBodyChunk):
(WebCore::BackgroundFetch::Record::didFinish):
(WebCore::BackgroundFetch::Record::retrieveResponse):
(WebCore::BackgroundFetch::Record::retrieveRecordResponseBody):
* Source/WebCore/workers/service/background-fetch/BackgroundFetch.h: Added.
(WebCore::BackgroundFetch::identifier const):
(WebCore::BackgroundFetch::Record::create):
(WebCore::BackgroundFetch::Record::setAsCompleted):
(WebCore::BackgroundFetch::Record::isCompleted const):
(WebCore::BackgroundFetch::Record::responseDataSize const):
(WebCore::BackgroundFetch::isActive const):
(WebCore::BackgroundFetch::origin const):
(WebCore::BackgroundFetch::downloadTotal const):
(WebCore::BackgroundFetch::uploadTotal const):
* Source/WebCore/workers/service/background-fetch/BackgroundFetchEngine.cpp: Added.
(WebCore::BackgroundFetchEngine::BackgroundFetchEngine):
(WebCore::BackgroundFetchEngine::startBackgroundFetch):
(WebCore::BackgroundFetchEngine::notifyBackgroundFetchUpdate):
(WebCore::BackgroundFetchEngine::backgroundFetchInformation):
(WebCore::BackgroundFetchEngine::backgroundFetchIdentifiers):
(WebCore::BackgroundFetchEngine::abortBackgroundFetch):
(WebCore::BackgroundFetchEngine::matchBackgroundFetch):
(WebCore::BackgroundFetchEngine::remove):
(WebCore::BackgroundFetchEngine::retrieveRecordResponse):
(WebCore::BackgroundFetchEngine::retrieveRecordResponseBody):
* Source/WebCore/workers/service/background-fetch/BackgroundFetchEngine.h: Added.
* Source/WebCore/workers/service/background-fetch/BackgroundFetchRecordLoader.h: Added.
* Source/WebCore/workers/service/background-fetch/BackgroundFetchStore.h: Added.
(WebCore::BackgroundFetchStore::clearRecords):
(WebCore::BackgroundFetchStore::clearAllRecords):
* Source/WebCore/workers/service/context/SWContextManager.cpp:
(WebCore::SWContextManager::fireBackgroundFetchEvent):
* Source/WebCore/workers/service/context/SWContextManager.h:
* Source/WebCore/workers/service/context/ServiceWorkerThread.cpp:
(WebCore::ServiceWorkerThread::queueTaskToFireBackgroundFetchEvent):
* Source/WebCore/workers/service/context/ServiceWorkerThread.h:
* Source/WebCore/workers/service/context/ServiceWorkerThreadProxy.cpp:
(WebCore::ServiceWorkerThreadProxy::fireBackgroundFetchEvent):
* Source/WebCore/workers/service/context/ServiceWorkerThreadProxy.h:
* Source/WebCore/workers/service/server/SWServer.cpp:
(WebCore::SWServer::removeRegistration):
(WebCore::SWServer::fireBackgroundFetchEvent):
(WebCore::SWServer::backgroundFetchEngine):
(WebCore::SWServer::Connection::startBackgroundFetch):
(WebCore::SWServer::Connection::backgroundFetchInformation):
(WebCore::SWServer::Connection::backgroundFetchIdentifiers):
(WebCore::SWServer::Connection::abortBackgroundFetch):
(WebCore::SWServer::Connection::matchBackgroundFetch):
(WebCore::SWServer::Connection::retrieveRecordResponse):
(WebCore::SWServer::Connection::retrieveRecordResponseBody):
* Source/WebCore/workers/service/server/SWServer.h:
(WebCore::SWServer::createBackgroundFetchRecordLoader):
(WebCore::SWServer::requestBackgroundFetchSpace):
(WebCore::SWServer::createBackgroundFetchStore):
* Source/WebCore/workers/service/server/SWServerDelegate.h:
* Source/WebCore/workers/service/server/SWServerToContextConnection.h:
* Source/WebKit/NetworkProcess/BackgroundFetchLoad.cpp: Added.
(WebKit::toSecurityOrigin):
(WebKit::BackgroundFetchLoad::BackgroundFetchLoad):
(WebKit::BackgroundFetchLoad::initialize):
(WebKit::BackgroundFetchLoad::~BackgroundFetchLoad):
(WebKit::BackgroundFetchLoad::abort):
(WebKit::BackgroundFetchLoad::didFinish):
(WebKit::BackgroundFetchLoad::loadRequest):
(WebKit::BackgroundFetchLoad::willPerformHTTPRedirection):
(WebKit::BackgroundFetchLoad::didReceiveChallenge):
(WebKit::BackgroundFetchLoad::didReceiveResponse):
(WebKit::BackgroundFetchLoad::didReceiveData):
(WebKit::BackgroundFetchLoad::didCompleteWithError):
(WebKit::BackgroundFetchLoad::didSendData):
(WebKit::BackgroundFetchLoad::wasBlocked):
(WebKit::BackgroundFetchLoad::cannotShowURL):
(WebKit::BackgroundFetchLoad::wasBlockedByRestrictions):
(WebKit::BackgroundFetchLoad::wasBlockedByDisabledFTP):
(WebKit::BackgroundFetchLoad::currentURL const):
* Source/WebKit/NetworkProcess/BackgroundFetchLoad.h: Added.
* Source/WebKit/NetworkProcess/NetworkLoadChecker.h:
(WebKit::NetworkLoadChecker::origin const):
(WebKit::NetworkLoadChecker::topOrigin const):
* Source/WebKit/NetworkProcess/NetworkResourceLoader.cpp:
* Source/WebKit/NetworkProcess/NetworkSession.cpp:
(WebKit::NetworkSession::createBackgroundFetchRecordLoader):
(WebKit::NetworkSession::requestBackgroundFetchSpace):
(WebKit::NetworkSession::createBackgroundFetchStore):
* Source/WebKit/NetworkProcess/NetworkSession.h:
* Source/WebKit/NetworkProcess/ServiceWorker/WebSWServerToContextConnection.cpp:
(WebKit::WebSWServerToContextConnection::fireBackgroundFetchEvent):
* Source/WebKit/NetworkProcess/ServiceWorker/WebSWServerToContextConnection.h:
* Source/WebKit/NetworkProcess/storage/BackgroundFetchStoreImpl.cpp: Added.
* Source/WebKit/NetworkProcess/storage/BackgroundFetchStoreImpl.h: Added.
* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
(WebKit::NetworkStorageManager::createBackgroundFetchStore):
* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.h:
* Source/WebKit/Sources.txt:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/Storage/WebSWContextManagerConnection.cpp:
(WebKit::WebSWContextManagerConnection::fireBackgroundFetchEvent):
* Source/WebKit/WebProcess/Storage/WebSWContextManagerConnection.h:
* Source/WebKit/WebProcess/Storage/WebSWContextManagerConnection.messages.in:

Canonical link: <a href="https://commits.webkit.org/260928@main">https://commits.webkit.org/260928@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3142406670cd5eb2406247efc9e8b982256a2d13

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/109991 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/19092 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/42674 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/1435 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/119038 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/113945 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/20557 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/10285 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/102249 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/115738 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/15304 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/98516 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/43522 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/97266 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/30173 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/85345 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/11794 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/31513 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/12417 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/8472 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/17785 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/62/builds/51124 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/7578 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/14212 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebCore/workers/service/WorkerSWClientConnection.cpp:
```diff
@@ -381,6 +410,148 @@ void WorkerSWClientConnection::getNavigationPreloadState(ServiceWorkerRegistrati
+    });
+}
+
+using CrossThreadResponseDataOrException = ExceptionOr<ResourceResponse::CrossThreadData>;
```
in header?  This isn't very useful.

youennf replied:
Removed

achristensen07 commented on Source/WebCore/workers/service/WorkerSWClientConnection.cpp:
```diff
@@ -381,6 +410,148 @@ void WorkerSWClientConnection::getNavigationPreloadState(ServiceWorkerRegistrati
 
+void WorkerSWClientConnection::startBackgroundFetch(ServiceWorkerRegistrationIdentifier registrationIdentifier, const String& backgroundFetchIdentifier, Vector<BackgroundFetchRequest>&& requests, BackgroundFetchOptions&& options, ExceptionOrBackgroundFetchInformationCallback&& callback)
+{
+    uint64_t requestIdentifier = ++m_lastRequestIdentifier;
```
Can these be strongly typed?

youennf replied:
We could but this is not strictly needed since this remains internal to the WorkerSWClientConnection method.

achristensen07 commented on Source/WebCore/workers/service/background-fetch/BackgroundFetchEngine.cpp:
```diff
@@ -0,0 +1,266 @@
+
+    // FIXME: We should wait for upload data to be fully read.
+    uint64_t expectedSpace;
+    constexpr uint64_t fixedRecordSize = 4096;
```
Is this based on anything?

youennf replied:
It was based on URL size but let's remove it for now, this is not strictly needed

achristensen07 commented on Source/WebKit/NetworkProcess/BackgroundFetchLoad.cpp:
```diff
@@ -0,0 +1,234 @@
+    , m_request(WTFMove(request))
+    , m_networkLoadChecker(makeUniqueRef<NetworkLoadChecker>(networkProcess, nullptr, nullptr, WTFMove(options), m_sessionID, WebPageProxyIdentifier { }, WebCore::HTTPHeaderMap { }, URL { m_request.url() }, URL { }, toSecurityOrigin(clientOrigin.clientOrigin), toSecurityOrigin(clientOrigin.topOrigin), RefPtr<SecurityOrigin> { }, PreflightPolicy::Consider, m_request.httpReferrer(), true, OptionSet ...
+{
+    initialize(networkProcess);
```
This seems undesirable.

youennf replied:
I inlined initialize

achristensen07 commented on Source/WebKit/NetworkProcess/BackgroundFetchLoad.cpp:
```diff
@@ -0,0 +1,234 @@
+    loadParameters.sourceOrigin = m_networkLoadChecker->topOrigin();
+//    loadParameters.storedCredentialsPolicy = == FetchOptions::Credentials::Include) ? StoredCredentialsPolicy::Use : StoredCredentialsPolicy::DoNotUse;
+
+    /*
```
Remove

youennf replied:
Done

achristensen07 commented on Source/WebKit/NetworkProcess/BackgroundFetchLoad.cpp:
```diff
@@ -0,0 +1,234 @@
+
+void BackgroundFetchLoad::willPerformHTTPRedirection(ResourceResponse&& redirectResponse, ResourceRequest&& request, RedirectCompletionHandler&& completionHandler)
+{
+    m_networkLoadChecker->checkRedirection(ResourceRequest { }, WTFMove(request), WTFMove(redirectResponse), nullptr, [this, completionHandler = WTFMove(completionHandler)] (auto&& result) mutable {
```
"this" is unsafe

youennf replied:
Switching to weakThis

achristensen07 reviewed (commented):
Needs feature flag, and spec links should probably be included in relevant places.

szewai commented on Source/WebCore/workers/service/background-fetch/BackgroundFetchEngine.cpp:
```diff
@@ -0,0 +1,270 @@
+
+    // FIXME: We should wait for upload data to be fully read.
+    uint64_t expectedSpace;
+    bool isSafe = WTF::safeAdd(fetch->downloadTotal(), fetch->uploadTotal(), expectedSpace);
```
Do we need to request space for downloadTotal? Is it possible some content is already on disk?

youennf replied:
The download body size is capped by downloadTotal.
And we might start the download when there is no longer a context of the given origin.
Hence why it seems good to preallocate space.

szewai commented on Source/WebCore/workers/service/background-fetch/BackgroundFetchEngine.h:
```diff
@@ -0,0 +1,73 @@
+    void notifyBackgroundFetchUpdate(BackgroundFetchInformation&&);
+
+    WeakPtr<SWServer> m_server;
+    Ref<BackgroundFetchStore> m_store;
```
It seems we can make it one `BackgroundFetchStore` per `ServiceWorkerRegistrationKey`, since all its functions take `ServiceWorkerRegistrationKey`.
Then we may merge it with the `m_fetches`: e.g. `HashMap<ServiceWorkerRegistrationKey, Fetches>`, and `Fetches` contains store.
This way we could make sure when `ServiceWorkerRegistrationKey` is gone (removed from `m_fetches`), we will remove all its objects, and we won't need additional keying in `BackgroundFetchStore`.
(Also by doing this we might store `ServiceWorkerRegistrationKey` in `BackgroundFetchStore` instead of `BackgroundFetch`)

youennf replied:
At WebKit layer, we need one object to handle all fetches (to handle pause/resume/click orders coming from UIProcess). The store is the one responsible for that.
We could add intermediary objects for each origin and/or each registration that would interact with that single object, but I am not sure this is worth it.
Also to handle SWServer::clear(SecurityOrigin), we might add a BackgroundFetchStore::clear(SecurityOrigin) since we might lazily load each origin.

szewai commented on Source/WebKit/NetworkProcess/storage/BackgroundFetchStoreImplementation.h:
```diff
@@ -0,0 +1,62 @@
+    void storeFetchResponseBodyChunk(const WebCore::ServiceWorkerRegistrationKey&, const String&, size_t, const WebCore::SharedBuffer&, CompletionHandler<void(StoreResult)>&&) final;
+    void retrieveResponseBody(const WebCore::ServiceWorkerRegistrationKey&, const String&, size_t, RetrieveRecordResponseBodyCallback&&) final;
+
+    Ref<NetworkStorageManager> m_manager;
```
Can we make `BackgroundFetchStoreImplementation` not hold strong ref to `NetworkStorageManager`? When `NetworkSession` is removed, it would close `NetworkStorageManager` (meaning `NetworkStorageManager` should not accept any new task); so I think this might be better hold weak ref to `NetworkSession` and ask it for `NetworkStorageManager`, or hold a weak ref to `NetworkStorageManager`.

youennf replied:
Yes, let's go with a WeakRef.

szewai commented on Source/WebKit/NetworkProcess/storage/BackgroundFetchStoreImplementation.h:
```diff
@@ -0,0 +1,62 @@
+    void retrieveResponseBody(const WebCore::ServiceWorkerRegistrationKey&, const String&, size_t, RetrieveRecordResponseBodyCallback&&) final;
+
+    Ref<NetworkStorageManager> m_manager;
+    Ref<WTF::WorkQueue> m_queue;
```
I think we probably should not expose the `WorkQueue` of `NetworkStorageManager` to another object that's not owned by it. As mentioned above, `NetworkStorageManager` should not accept new task when it's closed, and it might be easier for us to enforce the check inside `NetworkStorageManager`.

We could add new functions to `NetworkStorageManager` and let `NetworkStorageManager` do the task dispatch and close/thread check (like how `NetworkStorageManager` accept tasks on main thread nowadays).

youennf replied:
Good point, I'll better isolate things.

szewai commented on Source/WebKit/NetworkProcess/storage/BackgroundFetchStoreImplementation.h:
```diff
@@ -0,0 +1,62 @@
+
+class NetworkStorageManager;
+
+class BackgroundFetchStoreImplementation :  public WebCore::BackgroundFetchStore {
```
It seems we usually don't use the whole word `Implementation` in class name, so I would suggest renaming it `BackgroundFetchStoreImpl` or `WebBackgroundFetchStore`.

Also given that this is an implementation of ServiceWorker object and it does not actually implement storage functionalities (which I think would be implemented in `NetworkStorageManager`/`OriginStorageManager`/`BackgroundFetchStorageManager`), maybe this class should be moved to Source/WebKit/NetworkProcess/ServiceWorker/.

szewai commented on Source/WebKit/NetworkProcess/storage/NetworkStorageManager.h:
```diff
@@ -110,6 +111,10 @@ class NetworkStorageManager final : public IPC::WorkQueueMessageReceiver {
 #endif
 
+#if ENABLE(SERVICE_WORKER)
+    Ref<WebCore::BackgroundFetchStore> createBackgroundFetchStore();
```
I don't think we should add this function here for the reason I mentioned above. All `BackgroundFetchStore` or `BackgroundFetchStoreImplementation` needs to know is how to access `NetworkStorageManager` (via a weak ref to `NetworkStorageManager` or a weak ref to `NetworkSession`), and this can be done in `NetworkSession::createBackgroundFetchStore`. 

`NetworkStorageManager` does not need to know what `BackgroundFetchStore` is.

youennf replied:
I'll move it to NetworkSession level

szewai commented on Source/WebKit/NetworkProcess/BackgroundFetchLoad.cpp:
```diff
@@ -0,0 +1,224 @@
+    NetworkLoadParameters loadParameters;
+    loadParameters.request = WTFMove(request);
+    loadParameters.topOrigin = m_networkLoadChecker->topOrigin();
+    loadParameters.sourceOrigin = m_networkLoadChecker->topOrigin();
```
Is it expected that `sourceOrigin = topOrigin`?

youennf replied:
Right, will fix it

szewai commented on Source/WebCore/workers/service/background-fetch/BackgroundFetch.h:
```diff
@@ -0,0 +1,154 @@
+    bool m_recordsAvailableFlag { true };
+    ServiceWorkerRegistrationKey m_registrationKey;
+    ServiceWorkerRegistrationIdentifier m_registrationIdentifier;
+    bool m_abortFlag { false };
```
Can we name this `m_isAborted` and `m_areRecordsAvailable`? It seems more straightforward that it's a boolean.

youennf replied:
This is following the spec terminology.

szewai commented on Source/WebCore/workers/service/background-fetch/BackgroundFetch.h:
```diff
@@ -0,0 +1,154 @@
+    void handleStoreResult(BackgroundFetchStore::StoreResult);
+    void updateBackgroundFetchStatus(BackgroundFetchResult, BackgroundFetchFailureReason);
+
+    void store();
```
Can we rename this to `storeRecords()` or something else, since we have a member called `m_store` and `store()` feels like it will return `m_store`?

youennf replied:
We are storing more than records, let's go with doStore() then

szewai commented on Source/WebCore/workers/service/background-fetch/BackgroundFetchEngine.cpp:
```diff
@@ -0,0 +1,270 @@
+    if (iterator == m_fetches.end()) {
+        m_store->initializeFetches(*this, registration.key(), [weakThis = WeakPtr { *this }, registration = WeakPtr { registration }, backgroundFetchIdentifier, requests = WTFMove(requests), options = WTFMove(options), callback = WTFMove(callback)]() mutable {
+            if (!weakThis || !registration) {
+                callback(makeUnexpected(ExceptionData { NotSupportedError, "BackgroundFetchEngine is gone"_s }));
```
Not sure if this is temporary, but should we return `InvalidState` instead of `NotSupported` here?

youennf replied:
Yes, InvalidStateError is better.

cdumez commented on Source/WebCore/workers/service/background-fetch/BackgroundFetch.cpp:
```diff
@@ -0,0 +1,351 @@
+    m_abortFlag = true;
+    m_isActive = false;
+
+    for (auto& record : m_records)
```
Are we sure abort() won't mutate m_records while we iterate?

youennf replied:
This is currently guaranteed since m_records is never appending nor removing records outside the constructor.

cdumez commented on Source/WebCore/workers/service/background-fetch/BackgroundFetch.cpp:
```diff
@@ -0,0 +1,351 @@
+void BackgroundFetch::perform(const CreateLoaderCallback& createLoaderCallback)
+{
+    m_currentDownloadSize = 0;
+    for (auto& record : m_records)
```
ditto.

youennf replied:
This is currently guaranteed since m_records is never appending nor removing records outside the constructor.

cdumez commented on Source/WebCore/workers/service/background-fetch/BackgroundFetch.cpp:
```diff
@@ -0,0 +1,351 @@
+    if (!m_responseBodyCallbacks.isEmpty()) {
+        auto buffer = SharedBuffer::create(data.data(), data.size());
+        for (auto& callback : m_responseBodyCallbacks)
+            callback(RefPtr { buffer.ptr() });
```
`callback(buffer.copyRef())` should work I think

youennf replied:
Will switch buffer to RefPtr

cdumez commented on Source/WebCore/workers/service/background-fetch/BackgroundFetch.h:
```diff
@@ -0,0 +1,154 @@
+    using RetrieveRecordResponseBodyCallback = Function<void(Expected<RefPtr<SharedBuffer>, ResourceError>&&)>;
+    using CreateLoaderCallback = Function<std::unique_ptr<BackgroundFetchRecordLoader>(BackgroundFetchRecordLoader::Client&, ResourceRequest&&, FetchOptions&&, const ClientOrigin&)>;
+
+    class Record : public BackgroundFetchRecordLoader::Client, public RefCounted<Record> {
```
final?

youennf replied:
ok

cdumez commented on Source/WebCore/workers/service/background-fetch/BackgroundFetch.h:
```diff
@@ -0,0 +1,154 @@
+    void perform(const CreateLoaderCallback&);
+
+    bool isActive() const { return m_isActive; }
+    ClientOrigin origin() const { return m_origin; }
```
`const ClientOrigin&`

cdumez commented on Source/WebCore/workers/service/background-fetch/BackgroundFetch.h:
```diff
@@ -0,0 +1,154 @@
+    uint64_t m_uploadTotal { 0 };
+    BackgroundFetchResult m_result { BackgroundFetchResult::EmptyString };
+    BackgroundFetchFailureReason m_failureReason { BackgroundFetchFailureReason::EmptyString };
+    bool m_recordsAvailableFlag { true };
```
Would be nice to keep the enums and booleans together for better packing.

youennf replied:
ok

cdumez commented on Source/WebCore/workers/service/background-fetch/BackgroundFetch.h:
```diff
@@ -0,0 +1,154 @@
+    bool m_recordsAvailableFlag { true };
+    ServiceWorkerRegistrationKey m_registrationKey;
+    ServiceWorkerRegistrationIdentifier m_registrationIdentifier;
+    bool m_abortFlag { false };
```
Those for e.g.

cdumez commented on Source/WebCore/workers/service/background-fetch/BackgroundFetchEngine.cpp:
```diff
@@ -0,0 +1,270 @@
+        return;
+    }
+
+    auto fetch = WeakPtr { *result.iterator->value };
```
`WeakPtr fetch { *result.iterator->value };` would be more concise.

cdumez commented on Source/WebCore/workers/service/background-fetch/BackgroundFetchStore.h:
```diff
@@ -0,0 +1,69 @@
+struct BackgroundFetchRecordInformation;
+struct BackgroundFetchRequest;
+struct ExceptionData;
+class ResourceError;
```
We don't mix the struct and class forward declares. Should be first one then the other.

youennf replied:
ok

cdumez commented on Source/WebCore/workers/service/background-fetch/BackgroundFetchStore.h:
```diff
@@ -0,0 +1,69 @@
+    virtual void clearFetch(const ServiceWorkerRegistrationKey&, const String&, CompletionHandler<void()>&& = [] { }) = 0;
+    virtual void clearAllFetches(const ServiceWorkerRegistrationKey&, CompletionHandler<void()>&& = [] { }) = 0;
+
+    enum class StoreResult { OK, QuotaError, InternalError };
```
: uint8_t.

youennf replied:
ok

cdumez commented on Source/WebKit/NetworkProcess/BackgroundFetchLoad.cpp:
```diff
@@ -0,0 +1,224 @@
+
+using namespace WebCore;
+
+static Ref<SecurityOrigin> toSecurityOrigin(const SecurityOriginData& data)
```
Why can't we rely on SecurityOriginData::securityOrigin()?

youennf replied:
Will switch

cdumez commented on Source/WebKit/NetworkProcess/NetworkLoadChecker.h:
```diff
@@ -104,6 +104,11 @@ class NetworkLoadChecker : public CanMakeWeakPtr<NetworkLoadChecker> {
+    RefPtr<WebCore::SecurityOrigin> origin() const { return m_origin; }
+    RefPtr<WebCore::SecurityOrigin> topOrigin() const { return m_topOrigin; }
+
+    WebCore::FetchOptions options() const { return m_options; }
```
const WebCore::FetchOptions& probably.

youennf replied:
right

cdumez reviewed (approved):
LGTM

szewai commented on Source/WebKit/NetworkProcess/BackgroundFetchLoad.h:
```diff
@@ -0,0 +1,88 @@
+class NetworkConnectionToWebProcess;
+class NetworkLoadChecker;
+class NetworkProcess;
+class NetworkSchemeRegistry;
```
Where is this used?

youennf replied:
Will remove, this is a left over

szewai commented on Source/WebKit/NetworkProcess/BackgroundFetchLoad.h:
```diff
@@ -0,0 +1,88 @@
+
+namespace WebKit {
+
+class NetworkConnectionToWebProcess;
```
Where is this used?

youennf replied:
Will remove, this is a left over

szewai reviewed (approved):
r=me with nits

youennf committed:
Add a functional background fetch backend
https://bugs.webkit.org/show_bug.cgi?id=252657
rdar://problem/105720611

Reviewed by Chris Dumez and Sihui Liu.

Implement background fetch backend as follows:
- BackgroundFetchEngine is the class that is used by SWServer to implement the background fetch functionality
- BackgroundFetch represents a back ground fetch and has BackgroundFetch::Record(s).
- BackgroundFetchRecordLoader and its BackgroundFetchLoad implementation is about the actual network load.
- BackgroundFetchStore and its WebKit implementation BackgroundFetchStoreImpl allow storing background fetches.
BackgroundFetchStoreImpl is currently an empty shell.

Covered by rebased tests.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/background-fetch/abort.https.window-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/background-fetch/content-security-policy.https.window-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/background-fetch/fetch-uploads.https.window-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/background-fetch/fetch.https.window-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/background-fetch/get-ids.https.window-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/background-fetch/get.https.window-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/background-fetch/match.https.window-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/background-fetch/mixed-content-and-allowed-schemes.https.window-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/background-fetch/port-blocking.https.window-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/background-fetch/update-ui.https.window-expected.txt:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/workers/service/ServiceWorkerRegistrationKey.h:
* Source/WebCore/workers/service/background-fetch/BackgroundFetch.cpp: Added.
(WebCore::BackgroundFetch::BackgroundFetch):
(WebCore::BackgroundFetch::~BackgroundFetch):
(WebCore::BackgroundFetch::information const):
(WebCore::BackgroundFetch::match):
(WebCore::BackgroundFetch::abort):
(WebCore::BackgroundFetch::perform):
(WebCore::BackgroundFetch::storeResponse):
(WebCore::BackgroundFetch::storeResponseBodyChunk):
(WebCore::BackgroundFetch::didSendData):
(WebCore::BackgroundFetch::didFinishRecord):
(WebCore::BackgroundFetch::handleStoreResult):
(WebCore::BackgroundFetch::recordIsCompleted):
(WebCore::BackgroundFetch::updateBackgroundFetchStatus):
(WebCore::BackgroundFetch::unsetRecordsAvailableFlag):
(WebCore::BackgroundFetch::Record::Record):
(WebCore::BackgroundFetch::Record::~Record):
(WebCore::BackgroundFetch::Record::isMatching const):
(WebCore::BackgroundFetch::Record::information const):
(WebCore::BackgroundFetch::Record::complete):
(WebCore::BackgroundFetch::Record::abort):
(WebCore::BackgroundFetch::Record::didSendData):
(WebCore::BackgroundFetch::Record::didReceiveResponse):
(WebCore::BackgroundFetch::Record::didReceiveResponseBodyChunk):
(WebCore::BackgroundFetch::Record::didFinish):
(WebCore::BackgroundFetch::Record::retrieveResponse):
(WebCore::BackgroundFetch::Record::retrieveRecordResponseBody):
* Source/WebCore/workers/service/background-fetch/BackgroundFetch.h: Added.
(WebCore::BackgroundFetch::identifier const):
(WebCore::BackgroundFetch::Record::create):
(WebCore::BackgroundFetch::Record::setAsCompleted):
(WebCore::BackgroundFetch::Record::isCompleted const):
(WebCore::BackgroundFetch::Record::responseDataSize const):
(WebCore::BackgroundFetch::isActive const):
(WebCore::BackgroundFetch::origin const):
(WebCore::BackgroundFetch::downloadTotal const):
(WebCore::BackgroundFetch::uploadTotal const):
* Source/WebCore/workers/service/background-fetch/BackgroundFetchEngine.cpp: Added.
(WebCore::BackgroundFetchEngine::BackgroundFetchEngine):
(WebCore::BackgroundFetchEngine::startBackgroundFetch):
(WebCore::BackgroundFetchEngine::notifyBackgroundFetchUpdate):
(WebCore::BackgroundFetchEngine::backgroundFetchInformation):
(WebCore::BackgroundFetchEngine::backgroundFetchIdentifiers):
(WebCore::BackgroundFetchEngine::abortBackgroundFetch):
(WebCore::BackgroundFetchEngine::matchBackgroundFetch):
(WebCore::BackgroundFetchEngine::remove):
(WebCore::BackgroundFetchEngine::retrieveRecordResponse):
(WebCore::BackgroundFetchEngine::retrieveRecordResponseBody):
* Source/WebCore/workers/service/background-fetch/BackgroundFetchEngine.h: Added.
* Source/WebCore/workers/service/background-fetch/BackgroundFetchRecordLoader.h: Added.
* Source/WebCore/workers/service/background-fetch/BackgroundFetchStore.h: Added.
(WebCore::BackgroundFetchStore::clearRecords):
(WebCore::BackgroundFetchStore::clearAllRecords):
* Source/WebCore/workers/service/context/SWContextManager.cpp:
(WebCore::SWContextManager::fireBackgroundFetchEvent):
* Source/WebCore/workers/service/context/SWContextManager.h:
* Source/WebCore/workers/service/context/ServiceWorkerThread.cpp:
(WebCore::ServiceWorkerThread::queueTaskToFireBackgroundFetchEvent):
* Source/WebCore/workers/service/context/ServiceWorkerThread.h:
* Source/WebCore/workers/service/context/ServiceWorkerThreadProxy.cpp:
(WebCore::ServiceWorkerThreadProxy::fireBackgroundFetchEvent):
* Source/WebCore/workers/service/context/ServiceWorkerThreadProxy.h:
* Source/WebCore/workers/service/server/SWServer.cpp:
(WebCore::SWServer::removeRegistration):
(WebCore::SWServer::fireBackgroundFetchEvent):
(WebCore::SWServer::backgroundFetchEngine):
(WebCore::SWServer::Connection::startBackgroundFetch):
(WebCore::SWServer::Connection::backgroundFetchInformation):
(WebCore::SWServer::Connection::backgroundFetchIdentifiers):
(WebCore::SWServer::Connection::abortBackgroundFetch):
(WebCore::SWServer::Connection::matchBackgroundFetch):
(WebCore::SWServer::Connection::retrieveRecordResponse):
(WebCore::SWServer::Connection::retrieveRecordResponseBody):
* Source/WebCore/workers/service/server/SWServer.h:
(WebCore::SWServer::createBackgroundFetchRecordLoader):
(WebCore::SWServer::requestBackgroundFetchSpace):
(WebCore::SWServer::createBackgroundFetchStore):
* Source/WebCore/workers/service/server/SWServerDelegate.h:
* Source/WebCore/workers/service/server/SWServerToContextConnection.h:
* Source/WebKit/NetworkProcess/BackgroundFetchLoad.cpp: Added.
(WebKit::toSecurityOrigin):
(WebKit::BackgroundFetchLoad::BackgroundFetchLoad):
(WebKit::BackgroundFetchLoad::initialize):
(WebKit::BackgroundFetchLoad::~BackgroundFetchLoad):
(WebKit::BackgroundFetchLoad::abort):
(WebKit::BackgroundFetchLoad::didFinish):
(WebKit::BackgroundFetchLoad::loadRequest):
(WebKit::BackgroundFetchLoad::willPerformHTTPRedirection):
(WebKit::BackgroundFetchLoad::didReceiveChallenge):
(WebKit::BackgroundFetchLoad::didReceiveResponse):
(WebKit::BackgroundFetchLoad::didReceiveData):
(WebKit::BackgroundFetchLoad::didCompleteWithError):
(WebKit::BackgroundFetchLoad::didSendData):
(WebKit::BackgroundFetchLoad::wasBlocked):
(WebKit::BackgroundFetchLoad::cannotShowURL):
(WebKit::BackgroundFetchLoad::wasBlockedByRestrictions):
(WebKit::BackgroundFetchLoad::wasBlockedByDisabledFTP):
(WebKit::BackgroundFetchLoad::currentURL const):
* Source/WebKit/NetworkProcess/BackgroundFetchLoad.h: Added.
* Source/WebKit/NetworkProcess/NetworkLoadChecker.h:
(WebKit::NetworkLoadChecker::origin const):
(WebKit::NetworkLoadChecker::topOrigin const):
* Source/WebKit/NetworkProcess/NetworkResourceLoader.cpp:
* Source/WebKit/NetworkProcess/NetworkSession.cpp:
(WebKit::NetworkSession::createBackgroundFetchRecordLoader):
(WebKit::NetworkSession::requestBackgroundFetchSpace):
(WebKit::NetworkSession::createBackgroundFetchStore):
* Source/WebKit/NetworkProcess/NetworkSession.h:
* Source/WebKit/NetworkProcess/ServiceWorker/WebSWServerToContextConnection.cpp:
(WebKit::WebSWServerToContextConnection::fireBackgroundFetchEvent):
* Source/WebKit/NetworkProcess/ServiceWorker/WebSWServerToContextConnection.h:
* Source/WebKit/NetworkProcess/storage/BackgroundFetchStoreImpl.cpp: Added.
* Source/WebKit/NetworkProcess/storage/BackgroundFetchStoreImpl.h: Added.
* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
(WebKit::NetworkStorageManager::createBackgroundFetchStore):
* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.h:
* Source/WebKit/Sources.txt:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/Storage/WebSWContextManagerConnection.cpp:
(WebKit::WebSWContextManagerConnection::fireBackgroundFetchEvent):
* Source/WebKit/WebProcess/Storage/WebSWContextManagerConnection.h:
* Source/WebKit/WebProcess/Storage/WebSWContextManagerConnection.messages.in:

Canonical link: https://commits.webkit.org/260928@main
